### ISSUE TITLE
Z80 RC2014 v3.5.0-dev.101 boot-up shows sporadic "Console on Unit #1"

### ISSUE THREAD

**Issue by skullandbones** [wwarthen/RomWBW#485]
@wwarthen I have finally got around to playing with my Z80 RC2014 and Z180 SC126 boards again after 12 months or so.

I was previously using v3.4.0 and the earlier v3.4.0-dev releases.

Hopefully, I am using the correct build selection for my Z80 RC2014 because it seems v3.5.0 made some changes in this area.

I have built your v3.5.0-dev.101 but on my Z80 RC2014 I am seeing the following sporadic "Console on Unit #1" messages and the console does not work. I have to power cycle and sometimes the console works.

```
RomWBW HBIOS v3.5.0-dev.101, 2024-12-21

RCBus [RCZ80_std] Z80 @ 7.372MHz
0 MEM W/S, 1 I/O W/S, INT MODE 1, Z2 MMU
512KB ROM, 512KB RAM, HEAP=0x2418
ROM VERIFY: 00 00 00 00 PASS

CTC: IO=0x88 NOT PRESENT
LCD: IO=0xDA NOT PRESENT
SIO0: IO=0x80 SIO MODE=115200,8,N,1
SIO1: IO=0x82 SIO MODE=115200,8,N,1
DSRTC: MODE=STD IO=0xC0 Sat 2024-12-21 16:38:44 CHARGE=OFF
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=RCWDC IO=0x50 NOT PRESENT
IDE: IO=0x10 MODE=RC
IDE0: ATA 8-BIT LBA BLOCKS=0x0077F800 SIZE=3839MB
IDE1: NO MEDIA
PPIDE: IO=0x20 PPI NOT PRESENT
SD: MODE=PIO IO=0x69 DEVICES=1
SD0: NO MEDIA
CH0: IO=0x3E NOT PRESENT
CH1: IO=0x3C NOT PRESENT
FP: IO=0x00 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      SIO0:       RS-232            115200,8,N,1
Char 1      SIO1:       RS-232            115200,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       CompactFlash      3839MB,LBA
Disk 3      IDE1:       Hard Disk         --
Disk 4      SD0:        SD Card           --


RCBus [RCZ80_std] Boot Loader

Boot [H=Help]: 

  Console on Unit #1

RCBus [RCZ80_std] Boot Loader

Boot [H=Help]: 

  Console on Unit #1

RCBus [RCZ80_std] Boot Loader

Boot [H=Help]: 

  Console on Unit #1

RCBus [RCZ80_std] Boot Loader

Boot [H=Help]: 

  Console on Unit #1ü

```
I note from your documentation that pressing space twice requests control of the console. I did not press the space key or any keys.

My RC2014 has a Z80 SIO dual UART but only channel A (unit #0) is connected to a serial to USB adaptor to my computer. using minicom. The first sporadic "Console on Unit #1" immediately occurs when booting-up, no keys are being pressed on channel A (unit #0). Pressing keys (not sure which ones) seems to cause additional "Console on Unit #1" to be output.

I have to power cycle, and sometimes the console works fine (no "Console on Unit #1").

I wired up channel B (unit #1) to my computer (channel A (unit #0) was disconnected with no serial to USB adaptor) and then I observed "Console on Unit #0" on channel B (unit #1). Perhaps this is a hint of a logic error?

My Z180 SC126 does not have this issue.

Source/HBIOS/romldr.asm contains the console take over request source code. I guess that I could take a look.

Regards,
Dean

**Comment by skullandbones** [comment]
@wwarthen Another scenario

```
RomWBW HBIOS v3.5.0-dev.101, 2024-12-21

RCBus [RCZ80_std] Z80 @ 7.376MHz
0 MEM W/S, 1 I/O W/S, INT MODE 1, Z2 MMU
512KB ROM, 512KB RAM, HEAP=0x2414
ROM VERIFY: 00 00 00 00 PASS

CTC: IO=0x88 NOT PRESENT
LCD: IO=0xDA NOT PRESENT
SIO0: IO=0x80 SIO MODE=115200,8,N,1
SIO1: IO=0x82 SIO MODE=115200,8,N,1
DSRTC: MODE=STD IO=0xC0 Sun 2024-12-22 18:59:56 CHARGE=OFF
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=RCWDC IO=0x50 NOT PRESENT
IDE: IO=0x10 MODE=RC
IDE0: ATA 8-BIT LBA BLOCKS=0x0077F800 SIZE=3839MB
IDE1: NO MEDIA
PPIDE: IO=0x20 PPI NOT PRESENT
SD: MODE=PIO IO=0x69 DEVICES=1
SD0: NO MEDIA
CH0: IO=0x3E NOT PRESENT
CH1: IO=0x3C NOT PRESENT
FP: IO=0x00 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      SIO0:       RS-232            115200,8,N,1
Char 1      SIO1:       RS-232            115200,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       CompactFlash      3839MB,LBA
Disk 3      IDE1:       Hard Disk         --
Disk 4      SD0:        SD Card           --


RCBus [RCZ80_std] Boot Loader

Boot [H=Help]:   <== unit #0 pressing space space return keys

  L           - List ROM Applications
  D           - Device Inventory
  R           - Reboot System
  W           - RomWBW Configure
  I <u> [<c>] - Set Console Interface/Baud Rate
  V [<n>]     - View/Set HBIOS Diagnostic Verbosity
  <u>[.<s>]   - Boot Disk Unit/Slice

Boot [H=Help]:   <== unit #0 pressing space space return keys   

  L           - List ROM Applications
  D           - Device Inventory
  R           - Reboot System
  W           - RomWBW Configure
  I <u> [<c>] - Set Console Interface/Baud Rate
  V [<n>]     - View/Set HBIOS Diagnostic Verbosity
  <u>[.<s>]   - Boot Disk Unit/Slice

Boot [H=Help]:    <== unit #0 pressing space space return keys

  Console on Unit #1   <== Spurious switch over to unit #1
```
In this scenario, SIO channel B (unit #1) is not connected (no serial to USB adaptor was connected).

I guess the logic is somehow not properly tracking the console.

Note that my Z80 RC2104 is move reliable at the moment and the console works on each boot-up.

I am now running tests with both SIO A and B channels connected.

Regards,
Dean

**Comment by wwarthen** [comment]
Hi @skullandbones,

There were some recent changes to romldr.asm, so I took a few minutes to do some regression testing.  I have an RC2014 Z80 with Dual Serial (SIO) which replicates your setup.  I am not seeing any issues in my case.  I do not have a problem with the console being taken over incorrectly.  If I connect a second serial line, I can takeover the console by pressing space twice.  So, at the moment, I don't think this is a code issue.

In your problem scenarios, you have one of the serial ports disconnected.  I suspect that the alternate serial port is seeing enough garbage characters that it eventually sees 2 space characters.  The takeover function is coded to require seeing 2 space characters in a row specifially to avoid "false" takeovers like yours.  This approach almost always works well, but it definitely does not guarantee there won't be false takeovers and I have personally seen this occur on some other platforms.

I believe that, by happenstance, your hardware tends to generate space characters when the SIO port is disconnected.  Since the receive line of an unconnected port is floating, it may receive almost anything.  Very minor hardware changes can affect this.  You have a couple options for dealing with this.

You could build a custom rom with `AUTOCON`  set to `FALSE`.  This will disable the mechanism entirely.

Alternatively, if you want to retain the functionality, you could add a high value (maybe 100K) pull-up resistor to both of the receive signals on the Dual Serial board (RX and RXB on Spencer's schematic of the Dual Serial board).

Thanks, Wayne

**Comment by skullandbones** [comment]
@wwarthen I agree with you that the RX line of the Z80 SIO was suspected as being floating. On channel B (unit #1) I have added a 100K resistor to 5V to the RX line (TX pin of the serial to USB adaptor). The existing inline 1K resistor remains in place to prevent excessive current on the RX line.

With the 100K resistor in place with no serial to USB adaptor connector in place, then the spurious console switching events no longer occur. This matches your expectations.

I ran some tests using both serial ports and I observed the following:

**1. The 2 spaces do not need to be contiguous characters.**

With the console being on unit #0 with just the empty prompt being shown (this means no keys pressed), then type random characters into unit #1 and then include 1 space, followed by some more random characters and then a space, this causes the console to switch to unit #1.

I am guessing that you were expecting 2 contiguous space characters to trigger the console request but it seems that 2 non-contiguous spaces in a stream of random characters is sufficient to trigger the console request.

This means that as soon as the non-active console receives 2 space characters in total then a console request is made. Consequently, a serial port receiving garbage characters is likely to eventually trigger the console request.


**2. The console request cannot be aborted once armed**

I am not sure whether this observation is the intended behaviour.

With unit #0 being the active console, type a command such as X for the flash updater but don't press the return key.

On unit #1 type space space to request the console, but this is blocked by the partial command entry on unit #0 as expected.

On unit #1 press return to enter the flash updater. Enter q to exit.

But then the console is switched to unit 1.

I am wondering whether the request for console should be ignored until the partial command has been fully entered when the return key has been entered. Otherwise, the request for the console is armed and could remained armed for minutes or longer.

In other words, if the active console receives characters after the inactive console has sent the space space then perhaps the console request should be dropped.


**3. Are RX framing errors being checked?**

To randomly receive a space character is likely because the ASCII code is 0x20 meaning that a false start bit and a false data bit are needed. However, if RX framing errors are checked then a false stop bit would also be needed.

Hopefully, the source code is dropping any received characters that have framing errors.

Thanks,

Best regards,
Dean

**Comment by wwarthen** [comment]
Thank you very much @skullandbones.  Excellent observations.

> **1. The 2 spaces do not need to be contiguous characters.**

Definite bug.  I will address this.  This should really help reduce false takeovers.

> **2. The console request cannot be aborted once armed**

What's happening is that the space characters are being buffered in the receive queue of the alternate port.  When the console takeover goes back to polling the line, it sees the queued characters.  The queues should be flushed at the start of the polling.  I will look into this.

> **3. Are RX framing errors being checked?**

This is a function of the serial drivers and is generally not being checked.  Although this is a deficiency of the serial drivers, it is somewhat intentional due to performance and code space.  I will consider revisiting this, but probably not in the near future.

Thanks,

Wayne

**Comment by wwarthen** [comment]
Hi @skullandbones,

Your input was very valuable.

As of commit 5544275, I have addressed items 1 & 2 above.  Albeit a valid issue, item 3 is a bigger issue.  I don't plan to address this in the near future.

I will wait a bit for any comments, but will then close this issue.

Thanks, Wayne

**Comment by skullandbones** [comment]
Hi @wwarthen,

I have built and tested v3.5.0-dev.102 on my Z80 RC2014 and I can confirm that your fixes for issues 1 & 2 work OK.

Thank you for crediting me in your commit.

Best regards,
Dean

**Comment by wwarthen** [comment]
Thanks for confirming the changes @skullandbones.

I am going to close this issue now.

Thanks, Wayne
